src/peer_select.cc

```diff
@@ -99,6 +101,129 @@ operator <<(std::ostream &os, const PeerSelectionDumper &fsd)
     return os;
 }
 
+/// An ICP ping timeout service.
+/// Protects event.cc (which is designed to handle a few unrelated timeouts)
+/// from exposure to thousands of ping-related timeouts on busy proxies.
+class PeerSelectorPingMonitor
+{
+public:
+    /// registers the given selector to be notified about the IPC ping timeout
+    void monitor(PeerSelector *);
+
+    /// removes a PeerSelector from the waiting list
+    void forget(PeerSelector *);
+
+    /// \returns a (nil) registration of a non-waiting peer selector
+    WaitingPeerSelectorPosition npos() { return selectors.end(); }
+
+private:
+    static void NoteWaitOver(void *monitor);
+
+    void startWaiting();
+    void abortWaiting();
+    void noteWaitOver();
+
+    WaitingPeerSelectors selectors; ///< \see WaitingPeerSelectors
+};
+
+/// monitors all PeerSelector ICP ping timeouts
+static PeerSelectorPingMonitor &
+PingMonitor()
+{
+    static const auto Instance = new PeerSelectorPingMonitor();
+    return *Instance;
+}
+
+/* PeerSelectorPingMonitor */
+
+/// PeerSelectorPingMonitor::noteWaitOver() wrapper
+void
+PeerSelectorPingMonitor::NoteWaitOver(void *raw)
+{
+    assert(raw);
+    static_cast<PeerSelectorPingMonitor*>(raw)->noteWaitOver();
+}
+
+/// schedules a single event to represent all waiting selectors
+void
+PeerSelectorPingMonitor::startWaiting()
+{
+    assert(!selectors.empty());
+    const auto interval = tvSubDsec(current_time, selectors.begin()->first);
+    eventAdd("PeerSelectorPingMonitor::NoteWaitOver", &PeerSelectorPingMonitor::NoteWaitOver, this, interval, 0, false);
+}
+
+/// undoes an earlier startWaiting() call
+void
+PeerSelectorPingMonitor::abortWaiting()
+{
+    // our event may be already in the AsyncCallQueue but that is OK:
+    // such queued calls cannot accumulate, and we ignore any stale ones
+    eventDelete(&PeerSelectorPingMonitor::NoteWaitOver, nullptr);
+}
+
+/// calls back all ready PeerSelectors and continues to wait for others
+void
+PeerSelectorPingMonitor::noteWaitOver()
+{
+    while (!selectors.empty() && current_time >= selectors.begin()->first) {
+        const auto selector = selectors.begin()->second;
+        CallBack(selector->al, [selector,this] {
+            selector->ping.monitorRegistration = npos();
+            AsyncCall::Pointer callback = asyncCall(44, 4, "PeerSelector::HandlePingTimeout",
+                cbdataDialer(PeerSelector::HandlePingTimeout, selector));
+            ScheduleCallHere(callback);
+        });
+        selectors.erase(selectors.begin());
+    }
+
+    if (!selectors.empty()) {
+        // Since abortWaiting() is unreliable, we may have been awakened by a
```

Comment:
```C++
// Since abortWaiting() is unreliable, we may have been awakened by a
// stale event A after event B has been scheduled. Now we are going to
// schedule event C. Prevent event accumulation by deleting B (if any).
abortWaiting();

startWaiting();
```

> This description does not make sense to me.

I see no bugs in the description, but we should, of course, adjust it if something is unclear. I just do not know what adjustments are needed. Please suggest any changes you seem appropriate after reading the clarifications below.

> If an event has been scheduled then its objects have already been moved off the event list and onto the list of async Calls about to be 'dialed'.

"event scheduled" in the comment means "event given to eventAdd()". I think your interpretation of that word may be different.

> The only way to guarantee no duplicate event scheduling is for startWaiting() to call abortWaiting() itself.

Always calling abortWaiting() from startWaiting() would introduce a rather confusing and a little expensive extra abortWaiting() call in some cases. The current code avoids those extra calls (to the extent possible) without adding any state. This comment is the only "expense".

Please note that abortWaiting() cannot guarantee complete event deletion because eventDelete() and, hence, abortWaiting() itself is unreliable. The comment in question already says that.

> If duplicates matter that much, code it that way to avoid mistakes in the various callers logic.

The code is already written to tolerate stale events (because they are unavoidable). The point of this comment is not duplicate events (at least not in the direct sense of that word -- two equal events) or stale events. The point is event _accumulation_ (i.e. a continued growth in the number of added but not yet fired or deleted events) -- a potentially dangerous and often well-hidden hazzard. The code the comment in question details guarantees that we have no more than one event in the event.cc queue at any point in time (there could be a second/stale event in the async call queue but we cannot avoid that, and that does not result in _accumulation_).

A naive implementation would not add an abortWaiting() call in _this_ event-reaction method because the method is called when/because the previously scheduled event has been activated/fired and (as one might expect) there should be no other scheduled but undeleted/unfired events. Hence, scheduling a new event here would not result in event accumulation. The comment explains why that logic is flawed and why this (un-intuitive) abortWaiting() call was added to actually prevent that accumulation.

Do you see a way to adjust the comment to make sense to you given the explanation above?

Reply:
> The only way to guarantee no duplicate event scheduling is for startWaiting() to call abortWaiting() itself.

I think that Alex's explanation is correct. It would be unreasonable in unconditionally abortWaiting() before 
 startWaiting(), for example, when we add the first element and there is nothing to 'abort' yet. Similarly, sometimes we need to abortWaiting() without startWaiting() - when the container becomes empty. So we should carefully distinguish each of 3 cases, calling the method if/when needed. Two other places of abortWaiting()/startWaiting() invocation look trivial. This case is tricky - so we had to add the comment here.

Reply:
Perhapse a resetWaiting() helper method that does the pair in the right order to guarantee no mistake adding or removing from the event queue?
  We would have a triplet of start/reset/abort - with any given case calling only one of those methods to update the event state.

Reply:
Unfortunately, I do not think 'resetWaiting()' would simplify things. How this method would make a proper decision whether calling abortWaiting()/startWating()? Each context now has its own set of conditions - we cannot pass all of them into this new method.